### PR TITLE
chore(electric): Format known errors for easier reading and include hints to guide users towards resolution

### PR DESCRIPTION
**Pull request by alco**
There is a set of Electric's failure modes that we know of in advance. Some have to do with improper configuration, others are specific to a given runtime environment or the database in use.

To help users resolve those known issues on their own, this pull request adds a new module for outputting nicely-formatted errors with additional context and pointers for the users. It also adds explicit handling of a bunch of known errors.

A few examples:

### Server and client SSL configuration mismatch

```
23:19:16.864 pid=<0.380.0> [info] Starting ElectricSQL 0.8.1-12-g363f631 in logical_replication mode.
23:19:16.873 pid=<0.384.0> [debug] Loading flag string from env ELECTRIC_FEATURES
23:19:16.875 pid=<0.384.0> [debug] Got feature flag configuration %{proxy_ddlx_assign: false, proxy_ddlx_grant: false, proxy_ddlx_revoke: false, proxy_ddlx_unassign: false} with default value: false
23:19:16.910 pid=<0.381.0> [info] Running Electric.Plug.Router with Bandit 1.0.0 at :::5133 (http)
23:19:16.914 pid=<0.999.0> origin=postgres_1 [debug] Attempting to initialize postgres_1: postgres@localhost:54321
23:19:16.918 pid=<0.999.0> origin=postgres_1 [info] connect: %{database: ~c"electric", host: ~c"localhost", ipv6: false, port: 54321, ssl: false, timeout: 5000, username: ~c"postgres"}
Interactive Elixir (1.15.2) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> 23:19:16.945 pid=<0.999.0> origin=postgres_1 [error] Initialization of Postgres state failed with reason: {:error, :invalid_authorization_specification}.
▓ ┌────────────────────┐
▓ │  CONNECTION ERROR  │
▓ ┕━━━━━━━━━━━━━━━━━━━━┙
▓ 
▓ Failed to initialize Postgres state:
▓   {:error, :invalid_authorization_specification}
▓
▓ It appears that the database is configured to only accept connections
▓ encrypted with SSL. Make sure you configure Electric with DATABASE_REQUIRE_SSL=true.
23:19:16.946 pid=<0.999.0> origin=postgres_1 [info] schedule retry: 2000
```

### Postgres not configured with wal_level=logical
<details>
  <summary>Before</summary>

The following error and stacktrace would be printed out 3 times in succession due to Elixir's supervisor restarting the process a few times:

```
15:04:50.501 pid=<0.1298.0> origin=postgres_1 [info] Successfully initialized origin postgres_1 at extension version 
15:04:50.514 pid=<0.1303.0> [info] Starting Proxy server listening on port 65432
15:04:50.516 pid=<0.1304.0> pg_producer=postgres_1 [info] Starting Elixir.Electric.Postgres.Extension.SchemaCache for postgres_1
15:04:50.519 pid=<0.1304.0> pg_producer=postgres_1 [warning] SchemaCache "postgres_1" registered as the global instance
15:04:50.537 pid=<0.1307.0> [debug] Elixir.Electric.Replication.Postgres.LogicalReplicationProducer init:: publication: 'electric_publication', slot: 'electric_replication_out_electric'
15:04:50.537 pid=<0.1307.0> [info] Starting replication from postgres_1
15:04:50.537 pid=<0.1307.0> [info] Connection settings: %{database: ~c"electric", host: ~c"localhost", ipv6: false, password: ~c"******", port: 54321, replication: ~c"database", ssl: false, timeout: 5000, username: ~c"postgres"}
15:04:50.554 pid=<0.1307.0> [debug] Elixir.Electric.Replication.Postgres.Client: CREATE_REPLICATION_SLOT "electric_replication_out_electric" LOGICAL pgoutput NOEXPORT_SNAPSHOT
15:04:50.558 pid=<0.1298.0> origin=postgres_1 [error] GenServer #PID<0.1298.0> terminating
** (CaseClauseError) no case clause matching: {:error, {{:shutdown, {:failed_to_start_child, :postgres_producer, {{:case_clause, {:error, {:error, :error, "55000", :object_not_in_prerequisite_state, "logical decoding requires wal_level >= logical", [file: "logical.c", line: "113", routine: "CheckLogicalDecodingRequirements", severity: "ERROR"]}}}, [{Electric.Replication.Postgres.Client, :create_slot, 2, [file: ~c"lib/electric/replication/postgres/client.ex", line: 126]}, {Electric.Replication.Postgres.LogicalReplicationProducer, :init, 1, [file: ~c"lib/electric/replication/postgres/logical_replication_producer.ex", line: 93]}, {GenStage, :init, 1, [file: ~c"lib/gen_stage.ex", line: 1816]}, {:gen_server, :init_it, 2, [file: ~c"gen_server.erl", line: 851]}, {:gen_server, :init_it, 6, [file: ~c"gen_server.erl", line: 814]}, {:proc_lib, :init_p_do_apply, 3, [file: ~c"proc_lib.erl", line: 240]}]}}}, {:child, :undefined, :sup, {Electric.Replication.PostgresConnectorSup, :start_link, [[write_to_pg_mode: :logical_replication, origin: "postgres_1", producer: Electric.Replication.Postgres.LogicalReplicationProducer, connection: [replication: "database", ipv6: false, ssl: false, host: "localhost", port: 54321, database: "electric", username: "postgres", password: "password", timeout: 5000], replication: [electric_connection: [host: "host.docker.internal", port: 5433, dbname: "electric", connect_timeout: 5000]], proxy: [listen: [port: 65432, transport_options: [:inet6]], use_http_tunnel?: false, password: "password", log_level: :info]]]}, :temporary, false, :infinity, :supervisor, [Electric.Replication.PostgresConnectorSup]}}}
    (electric 0.8.1-4-gdc0008d) lib/electric/replication/postgres_manager.ex:108: Electric.Replication.PostgresConnectorMng.handle_continue/2
    (stdlib 4.3) gen_server.erl:1123: :gen_server.try_dispatch/4
    (stdlib 4.3) gen_server.erl:865: :gen_server.loop/7
    (stdlib 4.3) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Last message: {:continue, :establish_repl_conn}
State: %Electric.Replication.PostgresConnectorMng.State{status: :establishing_repl_conn, conn_config: %{database: ~c"electric", host: ~c"localhost", ipv6: false, password: ~c"password", port: 54321, ssl: false, timeout: 5000, username: ~c"postgres"}, repl_config: %{electric_connection: [host: "host.docker.internal", port: 5433, dbname: "electric", connect_timeout: 5000], publication: "electric_publication", slot: "electric_replication_out_electric", subscription: "postgres_1"}, backoff: {{:backoff, 1000, 10000, 1000, :normal, :undefined, :undefined}, nil}, origin: "postgres_1", config: [write_to_pg_mode: :logical_replication, origin: "postgres_1", producer: Electric.Replication.Postgres.LogicalReplicationProducer, connection: [replication: "database", ipv6: false, ssl: false, host: "localhost", port: 54321, database: "electric", username: "postgres", password: "password", timeout: 5000], replication: [electric_connection: [host: "host.docker.internal", port: 5433, dbname: "electric", connect_timeout: 5000]], proxy: [listen: [port: 65432, transport_options: [:inet6]], use_http_tunnel?: false, password: "password", log_level: :info]], pg_connector_sup_monitor: nil}
```

</details>

<details>
  <summary>After</summary>

```
15:03:29.001 pid=<0.975.0> origin=postgres_1 [info] Successfully initialized origin postgres_1 at extension version 20231206130400
15:03:29.016 pid=<0.980.0> [info] Starting Proxy server listening on port 65432
15:03:29.018 pid=<0.981.0> pg_producer=postgres_1 [info] Starting Elixir.Electric.Postgres.Extension.SchemaCache for postgres_1
15:03:29.022 pid=<0.981.0> pg_producer=postgres_1 [warning] SchemaCache "postgres_1" registered as the global instance
15:03:29.043 pid=<0.984.0> [debug] Elixir.Electric.Replication.Postgres.LogicalReplicationProducer init:: publication: 'electric_publication', slot: 'electric_replication_out_electric'
15:03:29.043 pid=<0.984.0> [info] Starting replication from postgres_1
15:03:29.043 pid=<0.984.0> [info] Connection settings: %{database: ~c"electric", host: ~c"localhost", ipv6: false, password: ~c"******", port: 54321, replication: ~c"database", ssl: false, timeout: 5000, username: ~c"postgres"}
15:03:29.059 pid=<0.984.0> [debug] Elixir.Electric.Replication.Postgres.Client: CREATE_REPLICATION_SLOT "electric_replication_out_electric" LOGICAL pgoutput NOEXPORT_SNAPSHOT
▓ ┌────────────────────────────────┐
▓ │  DATABASE CONFIGURATION ERROR  │
▓ ┕━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┙
▓ 
▓ Your Postgres database is not configured with wal_level=logical.
▓ 
▓ Visit https://electric-sql.com/docs/usage/installation/postgres
▓ to learn more about Electric's requirements for Postgres.

••• Shutting down •••

[os_mon] memory supervisor port (memsup): Erlang has closed
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
```

</details>

### Replication slot already in use

<details>
  <summary>Before</summary>

```
00:41:52.694 pid=<0.669.0> origin=postgres_1 [info] Successfully initialized origin postgres_1 at extension version 
00:41:52.708 pid=<0.674.0> [info] Starting Proxy server listening on port 65432
00:41:52.710 pid=<0.675.0> pg_producer=postgres_1 [info] Starting Elixir.Electric.Postgres.Extension.SchemaCache for postgres_1
00:41:52.714 pid=<0.675.0> pg_producer=postgres_1 [warning] SchemaCache "postgres_1" registered as the global instance
00:41:52.733 pid=<0.678.0> [debug] Elixir.Electric.Replication.Postgres.LogicalReplicationProducer init:: publication: 'electric_publication', slot: 'electric_replication_out_electric'
00:41:52.733 pid=<0.678.0> [info] Starting replication from postgres_1
00:41:52.734 pid=<0.678.0> [info] Connection settings: %{database: ~c"electric", host: ~c"localhost", ipv6: false, password: ~c"******", port: 54321, replication: ~c"database", ssl: false, timeout: 5000, username: ~c"postgres"}
00:41:52.750 pid=<0.678.0> [debug] Elixir.Electric.Replication.Postgres.Client: CREATE_REPLICATION_SLOT "electric_replication_out_electric" LOGICAL pgoutput NOEXPORT_SNAPSHOT
00:41:52.759 pid=<0.676.0> [debug] Starting SchemaLoader pg connection: [write_to_pg_mode: :direct_writes, origin: "postgres_1", producer: Electric.Replication.Postgres.LogicalReplicationProducer, connection: [replication: "database", ipv6: false, ssl: false, host: "localhost", port: 54321, database: "electric", username: "postgres", password: "password", timeout: 5000], replication: [electric_connection: [host: nil, port: 5433, dbname: "electric", connect_timeout: 5000]], proxy: [listen: [port: 65432, transport_options: [:inet6]], use_http_tunnel?: false, password: "password", log_level: :info]]
00:41:52.790 pid=<0.678.0> [debug] Elixir.Electric.Replication.Postgres.Client start_replication: slot: 'electric_replication_out_electric', publication: 'electric_publication'
00:41:52.793 pid=<0.669.0> origin=postgres_1 [error] Initialization of replication connection to Postgres failed with reason: replication slot "electric_replication_out_electric" is active for PID 371. Another instance of Electric appears to be connected to this database.
00:41:52.794 pid=<0.669.0> origin=postgres_1 [info] schedule retry: 2000
```

</details>

<details>
  <summary>After</summary>

```
15:09:48.938 pid=<0.982.0> [info] Starting replication from postgres_1
15:09:48.938 pid=<0.982.0> [info] Connection settings: %{database: ~c"electric", host: ~c"localhost", ipv6: false, password: ~c"******", port: 54321, replication: ~c"database", ssl: false, timeout: 5000, username: ~c"postgres"}
15:09:48.954 pid=<0.982.0> [debug] Elixir.Electric.Replication.Postgres.Client: CREATE_REPLICATION_SLOT "electric_replication_out_electric" LOGICAL pgoutput NOEXPORT_SNAPSHOT
15:09:48.959 pid=<0.980.0> [debug] Starting SchemaLoader pg connection: [write_to_pg_mode: :logical_replication, origin: "postgres_1", producer: Electric.Replication.Postgres.LogicalReplicationProducer, connection: [replication: "database", ipv6: false, ssl: false, host: "localhost", port: 54321, database: "electric", username: "postgres", password: "password", timeout: 5000], replication: [electric_connection: [host: "host.docker.internal", port: 5434, dbname: "electric", connect_timeout: 5000]], proxy: [listen: [port: 65433, transport_options: [:inet6]], use_http_tunnel?: false, password: "password", log_level: :info]]
15:09:48.990 pid=<0.982.0> [debug] Elixir.Electric.Replication.Postgres.Client start_replication: slot: 'electric_replication_out_electric', publication: 'electric_publication'
15:09:48.994 pid=<0.973.0> origin=postgres_1 [error] Initialization of PostgresConnectorSup failed with reason: {:bad_return_value, {:error, {:error, :error, "55006", :object_in_use, "replication slot \"electric_replication_out_electric\" is active for PID 59", [file: "slot.c", line: "446", routine: "ReplicationSlotAcquire", severity: "ERROR"]}}}.
15:09:48.995 pid=<0.980.0> [debug] Terminating idle db connection #PID<0.984.0>
▓ ┌────────────────────┐
▓ │  CONNECTION ERROR  │
▓ ┕━━━━━━━━━━━━━━━━━━━━┙
▓ 
▓ Failed to establish replication connection to Postgres:
▓   replication slot "electric_replication_out_electric" is active for PID 59
▓ 
▓ Another instance of Electric appears to be connected to this database.
15:09:48.997 pid=<0.973.0> origin=postgres_1 [info] schedule retry: 2000
```

</details>

### eaddrinuse (one of Electric's listening ports is already taken)

<details>
  <summary>Before</summary>

```
15:42:36.967 pid=<0.354.0> [info] Starting ElectricSQL 0.8.1-9-gfdcaee6 in logical_replication mode.
15:42:36.977 pid=<0.358.0> [debug] Loading flag string from env ELECTRIC_FEATURES
15:42:36.982 pid=<0.358.0> [debug] Got feature flag configuration %{proxy_ddlx_assign: false, proxy_ddlx_grant: false, proxy_ddlx_revoke: false, proxy_ddlx_unassign: false} with default value: false
15:42:37.023 pid=<0.355.0> [error] Running Electric.Plug.Router with Bandit 1.0.0-pre.14 at http failed, port already in use
15:42:37.028 pid=<0.44.0> [notice] Application electric exited: exited in: Electric.Application.start(:normal, [])
    ** (EXIT) an exception was raised:
        ** (FunctionClauseError) no function clause matching in Electric.Application.log_child_error/2
            (electric 0.8.1-9-gfdcaee6) lib/electric/application.ex:123: Electric.Application.log_child_error({Bandit, :root_http_router_listener}, {:shutdown, {:failed_to_start_child, :listener, :eaddrinuse}})
            (electric 0.8.1-9-gfdcaee6) lib/electric/application.ex:119: Electric.Application.log_supervisor_startup_error/1
            (kernel 8.5.4) application_master.erl:293: :application_master.start_it_old/4
15:42:37.038 pid=<0.44.0> [notice] Application nimble_parsec exited: :stopped
15:42:37.038 pid=<0.44.0> [notice] Application pg_protocol exited: :stopped
15:42:37.039 pid=<0.44.0> [notice] Application req exited: :stopped
15:42:37.039 pid=<0.44.0> [notice] Application finch exited: :stopped
15:42:37.039 pid=<0.44.0> [notice] Application nimble_options exited: :stopped
15:42:37.039 pid=<0.44.0> [notice] Application mint_web_socket exited: :stopped
15:42:37.039 pid=<0.44.0> [notice] Application mint exited: :stopped
15:42:37.039 pid=<0.44.0> [notice] Application castore exited: :stopped
15:42:37.040 pid=<0.44.0> [notice] Application bandit exited: :stopped
15:42:37.040 pid=<0.44.0> [notice] Application hpax exited: :stopped
15:42:37.040 pid=<0.44.0> [notice] Application websock exited: :stopped
15:42:37.040 pid=<0.44.0> [notice] Application plug exited: :stopped
15:42:37.041 pid=<0.44.0> [notice] Application plug_crypto exited: :stopped
15:42:37.041 pid=<0.44.0> [notice] Application mime exited: :stopped
15:42:37.041 pid=<0.44.0> [notice] Application eex exited: :stopped
15:42:37.041 pid=<0.44.0> [notice] Application thousand_island exited: :stopped
15:42:37.041 pid=<0.44.0> [notice] Application nimble_pool exited: :stopped
15:42:37.041 pid=<0.44.0> [notice] Application pg_query_ex exited: :stopped
15:42:37.044 pid=<0.44.0> [notice] Application tzdata exited: :stopped
15:42:37.044 pid=<0.44.0> [notice] Application hackney exited: :stopped
15:42:37.044 pid=<0.44.0> [notice] Application metrics exited: :stopped
15:42:37.044 pid=<0.44.0> [notice] Application parse_trans exited: :stopped
15:42:37.044 pid=<0.44.0> [notice] Application syntax_tools exited: :stopped
15:42:37.044 pid=<0.44.0> [notice] Application certifi exited: :stopped
15:42:37.044 pid=<0.44.0> [notice] Application mimerl exited: :stopped
15:42:37.044 pid=<0.44.0> [notice] Application idna exited: :stopped
15:42:37.044 pid=<0.44.0> [notice] Application unicode_util_compat exited: :stopped
15:42:37.045 pid=<0.44.0> [notice] Application exqlite exited: :stopped
15:42:37.045 pid=<0.44.0> [notice] Application db_connection exited: :stopped
15:42:37.045 pid=<0.44.0> [notice] Application stream_data exited: :stopped
15:42:37.045 pid=<0.44.0> [notice] Application pathex exited: :stopped
15:42:37.045 pid=<0.44.0> [notice] Application libgraph exited: :stopped
15:42:37.045 pid=<0.44.0> [notice] Application ets exited: :stopped
15:42:37.045 pid=<0.44.0> [notice] Application joken exited: :stopped
15:42:37.046 pid=<0.44.0> [notice] Application jose exited: :stopped
15:42:37.046 pid=<0.44.0> [notice] Application telemetry_metrics exited: :stopped
15:42:37.046 pid=<0.44.0> [notice] Application telemetry_poller exited: :stopped
15:42:37.047 pid=<0.44.0> [notice] Application telemetry exited: :stopped
15:42:37.047 pid=<0.44.0> [notice] Application gen_stage exited: :stopped
15:42:37.047 pid=<0.44.0> [notice] Application protox exited: :stopped
15:42:37.047 pid=<0.44.0> [notice] Application gproc exited: :stopped
15:42:37.047 pid=<0.44.0> [notice] Application jason exited: :stopped
15:42:37.047 pid=<0.44.0> [notice] Application decimal exited: :stopped
15:42:37.047 pid=<0.44.0> [notice] Application ssl_verify_fun exited: :stopped
15:42:37.048 pid=<0.44.0> [notice] Application mox exited: :stopped
15:42:37.048 pid=<0.44.0> [notice] Application backoff exited: :stopped
15:42:37.048 pid=<0.44.0> [notice] Application epgsql exited: :stopped
15:42:37.048 pid=<0.221.0> [notice]     :alarm_handler: {:clear, :system_memory_high_watermark}
15:42:37.049 pid=<0.44.0> [notice] Application os_mon exited: :stopped
15:42:37.049 pid=<0.44.0> [notice] Application sasl exited: :stopped
** (Mix) Could not start application electric: exited in: Electric.Application.start(:normal, [])
    ** (EXIT) an exception was raised:
        ** (FunctionClauseError) no function clause matching in Electric.Application.log_child_error/2
            (electric 0.8.1-9-gfdcaee6) lib/electric/application.ex:123: Electric.Application.log_child_error({Bandit, :root_http_router_listener}, {:shutdown, {:failed_to_start_child, :listener, :eaddrinuse}})
            (electric 0.8.1-9-gfdcaee6) lib/electric/application.ex:119: Electric.Application.log_supervisor_startup_error/1
            (kernel 8.5.4) application_master.erl:293: :application_master.start_it_old/4
```

</details>

<details>
  <summary>After</summary>

```
15:45:07.214 pid=<0.373.0> [info] Starting ElectricSQL 0.8.1-9-gfdcaee6 in logical_replication mode.
15:45:07.223 pid=<0.377.0> [debug] Loading flag string from env ELECTRIC_FEATURES
15:45:07.225 pid=<0.377.0> [debug] Got feature flag configuration %{proxy_ddlx_assign: false, proxy_ddlx_grant: false, proxy_ddlx_revoke: false, proxy_ddlx_unassign: false} with default value: false
15:45:07.257 pid=<0.374.0> [error] Running Electric.Plug.Router with Bandit 1.0.0-pre.14 at http failed, port already in use
▓ ┌────────────────────────┐
▓ │  INITIALISATION ERROR  │
▓ ┕━━━━━━━━━━━━━━━━━━━━━━━━┙
▓ 
▓ Failed to open a socket to listen for HTTP/WebSocket connections on port 5133.
▓ 
▓ Another instance of Electric or a different application is already listening on the same port.

••• Shutting down •••
```

</details>

### CODE REVIEW

File: components/electric/lib/electric/plug/proxy_websocket_plug.ex [hunk 1]
```diff
@@ -37,24 +37,17 @@ defmodule Electric.Plug.ProxyWebsocketPlug do
...
+    case Bandit.WebSocket.UpgradeValidation.validate_upgrade(conn) do
+      :ok ->
+        upgrade_adapter(
+          conn,
+          :websocket,
+          {Electric.Postgres.Proxy.WebsocketServer, websocket_opts, []}
+        )
+
+      {:error, reason} ->
+        Logger.debug("Clients WebSocket connection failed with reason: #{reason}")
```
Comment (magnetised) [inline components/electric/lib/electric/plug/proxy_websocket_plug.ex:49]:
```suggestion
        Logger.debug("Client WebSocket connection failed with reason: #{reason}")
```

File: components/electric/lib/electric/plug/satellite_websocket_plug.ex [hunk 2]
```diff
@@ -44,32 +44,38 @@ defmodule Electric.Plug.SatelliteWebsocketPlug do
...
         send_resp(conn, code, body)
     end
   end
 
   defp check_if_valid_upgrade(%Plug.Conn{} = conn) do
-    if Bandit.WebSocket.Handshake.valid_upgrade?(conn) do
-      {:ok, conn}
-    else
+    with {:error, reason} <- Bandit.WebSocket.UpgradeValidation.validate_upgrade(conn) do
+      Logger.debug("Clients WebSocket connection failed with reason: #{reason}")
```
Comment (magnetised) [inline components/electric/lib/electric/plug/satellite_websocket_plug.ex:53]:
```suggestion
      Logger.debug("Client WebSocket connection failed with reason: #{reason}")
```

**Review by magnetised — APPROVED** [review]
:heart: love this. have added suggestions to keep the `Client WebSocket...` messages consistent

**Comment by KyleAMathews** [comment]
😍 so useful!

**Pull request merged**